datapath: Enable N/S LB for overlapping pod CIDR

aspsk opened the pull request:
See the commit description.

<!-- Description of change -->

```release-note
Enable N/S LB for overlapping pod CIDR
```

aspsk committed:
bpf: fix tc_nodeport_lb4_nat_lb

The tc_nodeport_lb4_nat_lb might fail with

   --- FAIL: TestBPF/tc_nodeport_lb4_nat_lb.o (0.01s)
        bpf_test.go:172: loading collection: inlining global data: global constants must be in .rodata.config, but found reference to .bss

if SECLABEL is used inside lib/nodeport.h as an initializer for a variable.
Fix this by defining SECLABEL as an int for the test.

Signed-off-by: Anton Protopopov <aspsk@isovalent.com>

julianwiedmann commented on bpf/tests/tc_nodeport_lb4_nat_lb.c:
```diff
@@ -111,6 +111,10 @@ mock_ctx_redirect(const struct __sk_buff *ctx __maybe_unused,
+#ifndef SECLABEL
+#define SECLABEL 0xffff
+#endif
+
```
We probably just missed to include `config_replacement.h` here (or didn't need it until now?)

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2706,7 +2709,7 @@ int tail_nodeport_nat_egress_ipv4(struct __ctx_buff *ctx)
 						IPV4_DIRECT_ROUTING,
 						src_port,
 						tunnel_endpoint,
-						WORLD_IPV4_ID,
+						src_identity,
```
Two blocking concerns here.

1. how do we preserve backwards compatibility for existing network policies?
2. we mustn't pass CIDR identities to other nodes, they have node-local scope. See https://github.com/cilium/cilium/pull/28441 :).

brb replied:
Do we have a helper function to check whether sec id is a CIDR one?

julianwiedmann replied:
> Do we have a helper function to check whether sec id is a CIDR one?

I suspect it's this:
https://github.com/cilium/cilium/blob/0bf036effa18390c96c188be00e19cdd923050e4/bpf/lib/identity.h#L142-L143

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2647,6 +2648,8 @@ int tail_nodeport_nat_egress_ipv4(struct __ctx_buff *ctx)
 	__s8 ext_err = 0;
 #ifdef TUNNEL_MODE
+	__u8 cluster_id __maybe_unused = (__u8)ctx_load_meta(ctx, CB_CLUSTER_ID_EGRESS);
+	__u32 src_identity = ctx_load_meta(ctx, CB_SRC_LABEL) ?: WORLD_IPV4_ID;
```
I think we should do this selection in `nodeport_lb4()` already. Also note that we currently use `WORLD_IPV4_ID` when [creating](https://github.com/cilium/cilium/blob/fa0037638534635471c8e45583d78f87ad78b3e1/bpf/lib/nodeport.h#L2917) the nodeport CT entry, should probably stay in sync (even though I have no clue what it's used for ...).

julianwiedmann reviewed (changes requested):
Few comments below, before I forget them ...

aspsk committed:
datapath: Enable N/S LB for overlapping pod CIDR

This commit enables access to a K8s service from outside (N/S)
when a request first comes to a node which belongs to cluster which does
NOT run a service backend, and the overlapping pod CIDR feature is
enabled. For example:

  ┌───────────────────────┐
  │                       │
  │  Client from outside  │
  │                       │
  └──┬────────────────────┘
     │
     │ curl $NODE1:$NODEPORT
     │
  ┌──▼─────────────────────┐
  │                        │
  │  K8s Node1@Cluster1    │
  │                        │
  └──┬─────────────────────┘
     │
     │ SNAT + vxlan
     │ (overlapping pod CIDR)
     │
  ┌──▼─────────────────────┐
  │                        │
  │  K8s Node2@Cluster2    │
  │                        │
  └────────────────────────┘

To make it possible, the N/S LB entry function "nodeport_lb4()" is modified
to pass a dst cluster ID to "tail_nodeport_nat_egress_ipv4()". This
allows the latter to query a correct service endpoint in the IPcache,
and then to set an SNAT IP addr accordingly.

Co-authored-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Anton Protopopov <aspsk@isovalent.com>
